### PR TITLE
feat:  [v23 port]  data contract indices validation

### PR DESCRIPTION
**Pull request by qrayven**
The change includes: 
  -  validator of data contract indices backward compatibility with included changes from https://github.com/dashevo/platform/pull/52

### CODE REVIEW

File: dpp/src/errors/codes.rs [hunk 1]
```diff
@@ -105,6 +105,10 @@ impl ErrorWithCode for BasicError {
             Self::InvalidJsonSchemaRefError { .. } => 1014,
             Self::InconsistentCompoundIndexDataError { .. } => 1021,
 
+            Self::DataContractUniqueIndicesChangedError { .. } => 4016,
+            // TODO  -  they don't have error codes in  https://github.com/dashevo/platform/blob/25ab6d8a38880eaff6ac119126b2ee5991b2a5aa/packages/js-dpp/lib/errors/consensus/codes.js
+            Self::DataContractHaveNewUniqueIndexError { .. } => 0,
```
Comment (antouhou) [inline dpp/src/errors/codes.rs:111]: @shumkov @shuplenkov @jawid-h Do you guys know anything about why those errors do not have codes?

**Review by antouhou — DISMISSED** [review]
Great job 👍 I would wait a little but before guys from platform team comments on the error codes though

File: dpp/src/errors/codes.rs (hunk 1 above)
Comment (shumkov) [reply dpp/src/errors/codes.rs:111]: I would recommend comparing it with v0.23. I believe we fixed it.

File: dpp/src/util/json_schema.rs [hunk 2]
```diff
@@ -87,34 +88,25 @@ impl JsonSchemaExt for JsonValue {
 
     fn get_indices(&self) -> Result<Vec<Index>, anyhow::Error> {
         match self.get("indices") {
-            Some(raw_indices) => Ok(serde_json::from_value(raw_indices.to_owned())?),
+            Some(raw_indices) => {
+                let indices_with_raw_properties: Vec<IndexWithRawProperties> =
```
Comment (shumkov) [inline dpp/src/util/json_schema.rs:92]: Nor sure about usage. But I guess we have new RS Drive primitives to get this info.

File: dpp/src/tests/data_contract/validation/basic/validate_indices_are_backward_compatible_spec.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+#[test]
+fn should_return_invalid_result_if_one_of_new_indices_is_unique() {
+    let TestData {
+        old_documents_schema,
+        mut new_documents_schema,
+        ..
+    } = setup_test();
+    new_documents_schema.get_mut("indexedDocument").unwrap()["indices"]
```
Comment (shumkov) [inline dpp/src/tests/data_contract/validation/basic/validate_indices_are_backward_compatible_spec.rs:204]: I would spit into twp variables for better readability

File: dpp/src/tests/data_contract/validation/basic/validate_indices_are_backward_compatible_spec.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+#[test]
+fn should_return_invalid_result_if_non_unique_index_update_failed_due_old_properties_used() {
+    let TestData {
+        old_documents_schema,
+        mut new_documents_schema,
+        ..
+    } = setup_test();
+    new_documents_schema.get_mut("indexedDocument").unwrap()["indices"][2]["properties"][0] =
```
Comment (shumkov) [inline dpp/src/tests/data_contract/validation/basic/validate_indices_are_backward_compatible_spec.rs:140]: I would spit into twp variables for better readability

**Review by shumkov — DISMISSED** [review]

**Comment by QuantumExplorer** [comment]
Will fix issues that Ivan pointed out after merge.

**Pull request merged**